crates/oracle/src/subgraph.rs

```diff
@@ -119,48 +118,28 @@ where
 {
     pub fn new(api: A) -> Self {
         Self {
-            last_state: None,
-            error: None,
+            last_result: Ok(None),
             subgraph_api: api,
         }
     }
 
-    pub fn is_valid(&self) -> bool {
-        self.error.is_none() && self.last_state.is_some()
-    }
-
-    pub fn is_uninitialized(&self) -> bool {
-        self.last_state.is_none()
-    }
-
-    pub fn is_failed(&self) -> bool {
-        self.error.is_some() && self.last_state.is_some()
-    }
-
-    pub fn last_state(&self) -> Option<&A::State> {
-        self.last_state.as_ref()
-    }
-
-    pub fn error(&self) -> Option<Arc<A::Error>> {
-        self.error.clone()
+    pub fn result(&self) -> Result<Option<&A::State>, Arc<A::Error>> {
+        match self.last_result {
+            Ok(Some(ref s)) => Ok(Some(s)),
+            Ok(None) => Ok(None),
+            Err(ref e) => Err(e.clone()),
+        }
     }
 
     /// Handles the retrieval of new subgraph state and the transition of its internal [`State`]
     pub async fn refresh(&mut self) {
         info!("Fetching latest subgraph state");
 
-        match self.subgraph_api.get_subgraph_state().await {
-            Ok(s) => {
-                self.last_state = s;
-                self.error = None;
-            }
-            Err(err) => {
-                if self.is_failed() {
-                    error!("Failed to retrieve state from a previously failed subgraph");
-                }
-                self.error = Some(Arc::new(err));
-            }
-        }
+        self.last_result = self
+            .subgraph_api
+            .get_subgraph_state()
+            .await
+            .map_err(Arc::new);
```

Comment:
While we can get rid of state retention, I think we could seize this moment of transition to check if the previous state was altered or not. I believe this could be useful from a monitoring and alerting perspective.